go/genkit/chat.go

```diff
@@ -0,0 +1,335 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package genkit
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/firebase/genkit/go/ai"
+)
+
+type Chat struct {
+	// Genkit
+	Genkit *Genkit `json:"genkit,omitempty"`
+	// The model to query
+	Model ai.Model `json:"model,omitempty"`
+	// The chats threadname
+	ThreadName string `json:"threadName,omitempty"`
+	// The chats session
+	Session *Session `json:"session,omitempty"`
+	// Message sent to the model as system instructions
+	SystemText string `json:"systemtext,omitempty"`
+	// Optional prompt
+	Prompt *ai.Prompt `json:"prompt,omitempty"`
+	// Optional input fields for the chat. This should be a struct,
+	// a pointer to a struct that matches the input schema, or a string.
+	Input any `json:"input,omitempty"`
+
+	Request *ChatRequest              `json:"request,omitempty"`
+	Stream  ai.ModelStreamingCallback `json:"stream,omitempty"`
+}
+
+type ChatRequest struct {
+	Config   any           `json:"config,omitempty"`
+	Context  []any         `json:"context,omitempty"`
+	Messages []*ai.Message `json:"messages,omitempty"`
+
+	// Defines the output format and schema
+	Schema any             `json:"schema,omitempty"`
+	Format ai.OutputFormat `json:"outputformat,omitempty"`
+
+	// Tools lists the available tools that the model can ask the client to run.
+	Tools []ai.Tool `json:"tools,omitempty"`
+}
+
+// ChatOption configures params for the chat
+type ChatOption func(c *Chat) error
+
+// NewChat creates a new chat instance with the provided AI model and options.
+// If no session or thread name is provided, it adds a new session and default thread.
+func NewChat(ctx context.Context, g *Genkit, opts ...ChatOption) (chat *Chat, err error) {
+	chat = &Chat{
+		Genkit:  g,
+		Request: &ChatRequest{},
+	}
+
+	for _, with := range opts {
+		err := with(chat)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if chat.Session == nil {
+		s, err := NewSession(ctx)
+		if err != nil {
+			return nil, err
+		}
+		chat.Session = s
+	}
+
+	if chat.ThreadName == "" {
+		chat.ThreadName = "default"
+	}
+
+	return chat, nil
+}
+
+// Send sends a message to the chat, generating a response from the AI model.
+// It retrieves the chat history from the session store, adds the new message
+// to the history, and sends the entire conversation to the AI model for
+// generating a response. If a system message is set for the chat, it is
+// included in the conversation before the history.
+func (c *Chat) Send(ctx context.Context, msg any) (resp *ai.ModelResponse, err error) {
```

Comment:
I think we should break this out into two methods:

```go
func (c *Chat) Send(ctx context.Context, msgs ...*Message) (resp *ai.ModelResponse, err error)
func (c *Chat) SendText(ctx context.Context, msgText string) (resp *ai.ModelResponse, err error)
```

This is a very high touch method so maintaining strong typing here is essential. This should cover the majority of user journeys since even in the non-string scenarios we still convert to a text message.